rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,392 @@
+
+```go
+// AttestationResponse is veriried attestation data for a public key.
+type AttestationResponse struct {
+	// PublicKeyDER is the public key in PKIX, ASN.1 DER form.
+	PublicKeyDER []byte `json:"public_key"`
+	// PrivateKeyPolicy specifies the private key policy supported by the attested hardware key.
+	PrivateKeyPolicy PrivateKeyPolicy `json:"private_key_policy"`
+}
+```
+
+This `AttestationResponse` will then be checked against the user's private key policy requirement. If the policy requirement is met, the Auth server will sign the user's certificates with a private key policy extension matching the attestation. 
+
+```go
+// tls extension
+PrivateKeyPolicyASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 1, 15}
+
+// ssh extension
+CertExtensionPrivateKeyPolicy = "private-key-policy"
+```
+
+The `AttestationResponse` will also be stored in the backend under `/key_attestations/<sha256>` so that reissue requests can pass the attestation check without re-providing the attestation request. Key attestations will expire at the same time as the initial login certificates.
+
+#### Certificate key policy extension enforcement
+
+On every Teleport request that enforces valid certificates, we will check that the required private key policy extension is included. This check will be handled by Teleport's shared authorizer, in a similar way to [user locking](https://github.com/gravitational/teleport/blob/master/rfd/0009-locking.md) enforcement.
+
+### Client changes
+
+Teleport clients will need the ability to connect to a user's hardware key, generate/retrieve private keys, and use those keys for cryptographical operations. 
+
+#### Private key policy discovery
+
+Teleport clients should be able to automatically determine if a user requires a hardware private key for login to avoid additional UX concerns. Since it is not possible to retrieve a user's actual private key policy requirement before login, Teleport clients will make a best effort attempt to guess the key policy requirement.
+
+First, the client will ping the Teleport Auth server to get the cluster-wide private key policy if set. Second, the client will check for an existing key in the user's key store (`~/.tsh`), and check its associated private key policy. Between the two private key policies retrieved, the stricter one will be used for initial login. This guessing logic will capture all cases except for the case where a user's role private key policy is stricter than the cluster-wide policy, and do not have an active/expired login session stored in `~/.tsh`.
+
+If the private key policy was incorrect and a stricter requirement is neeeded, then the server will respond with a `private key policy not met: <private-key-policy>` error. The client will parse this error and resort to re-authenticating with the correct private key policy, meaning that the user will be re-prompted for their login credentials.
```

Comment:
> The client will parse this error

Let's try avoid parsing/matching strings if we can, or at least get additional signals: a specific HTTP error code for REST (plus the string?) or error details for gRPC.

* https://pkg.go.dev/google.golang.org/grpc/status#FromProto
* https://pkg.go.dev/google.golang.org/genproto/googleapis/rpc/status#Status

Reply:
I think that's for the best. We don't need to rush it, like you said, but I'd rather not drop it either.